design-docs/incremental-build.md

```diff
@@ -243,3 +243,115 @@ When the `TaskHistory` gets persisted, it adds the current task execution to the
+if the build changes the Gradle version or if `.gradle` is deleted manually _and_ one or many of the inputs have been changed. Any tasks using those outputs as inputs for further processing are affected as
+well.
+
+The issue is documented by [issue #821](https://github.com/gradle/gradle/issues/821). This story only addresses stale output files created by Java compilation.
+
+**Example:**
+
+A user compiles two Java source files, A and B, with Gradle version X via `compileJava`. The produced output are the class files C and D. Those class files might be packaged into
+ a JAR file via the `jar` task.
+
+Let's assume the user upgrades the version of Gradle to Y _and_ one of the input files is moved or deleted e.g. B is deleted (which might happen upon a branch change or refactoring of the code). 
+Gradle does not have a task history for the task `compileJava` as `.gradle` directory is based on the used Gradle version e.g. `./gradle/3.1` vs. `.gradle/3.2`. As a result `compileJava` would compile
+source file A but not B as it doesn't exist anymore. The output directory of the compilation task would still contain A and B which leads to an incorrect result.
+
+The goal of this story is to 1) detect this situation and 2) clean up any stale output.
+
+### User visible changes
+
+Gradle detects and removes stale Java class files after a Gradle version change. Further processing of output files (e.g. by the `jar` task) does not result in faulty behavior.
+
+### Implementation
+
+- Establish a registry implementation that allows for registering strategies with the purpose of cleaning up outputs produced by previous builds.
+    - The registry is not specific to tasks. It's rather a global concept that allows for registering clean up strategies for any "thing" in Gradle.
+    - The registry is going to be an internal concept and is not going to be exposed via a public API.
+    - The registry needs to be instantiated by Gradle's global service registry. 
+    - When a build is executed, the registry is contacted and provides all registered strategies.
+    - If any of the strategies indicate that a clean up is needed then delete all registered outputs.
+
+<!-- -->
+
+    public interface BuildOutputCleanupStrategy {
+        /**
+         * Determines if cleanup is needed based on provided Gradle instance.
+         */
+        boolean needsCleanup(Gradle gradle);
+    }
+
+    public interface BuildOutputCleanupRegistry {
+        /**
+         * Registers clean up strategy.
+         */
+        void registerStrategy(BuildOutputCleanupStrategy strategy);
+        
+        /**
+         * Returns all registered clean up strategies.
+         */
+        List<BuildOutputCleanupStrategy> getStrategies();
+        
+        /**
+         * Registers outputs to be cleaned up.
+         */
+        void registerOutputs(File... outputs);
```

Comment:
Right, this could be files or directories. I'd build in handling for the logic that deletes the outputs:

1) If output is a file, then delete the file.
2) If output is a directory, then clean the directory but do not delete it.